services/src/main/java/edu/unc/lib/dl/cdr/services/processing/AddContainerService.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.jena.rdf.model.ModelFactory;
+import org.apache.jena.rdf.model.Resource;
+
+import edu.unc.lib.dl.acl.service.AccessControlService;
+import edu.unc.lib.dl.acl.util.AgentPrincipals;
+import edu.unc.lib.dl.acl.util.Permission;
+import edu.unc.lib.dl.fcrepo4.ContentContainerObject;
+import edu.unc.lib.dl.fcrepo4.RepositoryObjectFactory;
+import edu.unc.lib.dl.fedora.PID;
+import edu.unc.lib.dl.rdf.Cdr;
+import edu.unc.lib.dl.rdf.Premis;
+
+/**
+ * Service that manages the creation of containers
+ *
+ * @author harring
+ *
+ */
+public class AddContainerService {
+
+    private AccessControlService aclService;
+    private RepositoryObjectFactory repoObjFactory;
+
+    /**
+     * Mark each pid for deletion using the agent principals provided.
+     *
+     * @param agent security principals of the agent making request.
+     * @param ids ids of objects to mark for deletion
+     */
+    public void addContainer(AgentPrincipals agent, PID parentPid, Resource containerType) {
+
+     // Create the new container
+        Model model = ModelFactory.createDefaultModel();
+        Resource parentResc = model.getResource(parentPid.getRepositoryPath());
+        ContentContainerObject child = null;
+        // Create the appropriate container
+        if (Cdr.AdminUnit.equals(containerType)) {
+            aclService.assertHasAccess(
+                    "User does not have permissions to create admin units",
+                    parentPid, agent.getPrincipals(), Permission.createAdminUnit);
+
+            child = repoObjFactory.createAdminUnit(model);
+            repoObjFactory.createMemberLink(URI.create(parentResc.getURI()), child.getPid().getRepositoryUri());
+        } else if (Cdr.Collection.equals(containerType)) {
+            aclService.assertHasAccess(
+                    "User does not have permissions to create collections",
+                    parentPid, agent.getPrincipals(), Permission.createCollection);
+
+            child = repoObjFactory.createCollectionObject(model);
+            repoObjFactory.createMemberLink(URI.create(parentResc.getURI()), child.getPid().getRepositoryUri());
+        } else if (Cdr.Folder.equals(containerType)) {
+            aclService.assertHasAccess(
+                    "User does not have permissions to create collections",
+                    parentPid, agent.getPrincipals(), Permission.ingest);
+
+            child = repoObjFactory.createFolderObject(model);
+            repoObjFactory.createMemberLink(URI.create(parentResc.getURI()), child.getPid().getRepositoryUri());
+        } else if (Cdr.Work.equals(containerType)) {
+            aclService.assertHasAccess(
+                    "User does not have permissions to create collections",
+                    parentPid, agent.getPrincipals(), Permission.ingest);
+
+            child = repoObjFactory.createWorkObject(model);
+            repoObjFactory.createMemberLink(URI.create(parentResc.getURI()), child.getPid().getRepositoryUri());
```

Comment:
I'd suggest doing this by getting the parentObject and calling addMember on it, since that includes validation that its a legitimate child type being added.  You could probably move this out of the ifs since its the same.  Should also make sure that the child was actually created (was a valid type) before continuing.  Might be best to to perform the creation and adding (and maybe the premis) in a transaction so you could back the creation of the object out if its an invalid child type (like adding an AdminUnit to a folder)